cmd/cosign/cli/upload_blob.go

```diff
@@ -0,0 +1,87 @@
+//
+// Copyright 2021 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cli
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/types"
+	"github.com/peterbourgon/ff/v3/ffcli"
+	cremote "github.com/sigstore/cosign/pkg/cosign/remote"
+)
+
+func UploadBlob() *ffcli.Command {
+	var (
+		flagset = flag.NewFlagSet("cosign upload-blob", flag.ExitOnError)
+		fileRef = flagset.String("f", "", "reference to the file to upload")
+		ct      = flagset.String("ct", "", "content type to set")
+	)
+	return &ffcli.Command{
+		Name:       "upload-blob",
+		ShortUsage: "cosign upload-blob -f <blob ref> <image uri>",
+		ShortHelp:  "upload a blob to the supplied container image",
+		FlagSet:    flagset,
+		Exec: func(ctx context.Context, args []string) error {
+			if len(args) != 1 {
+				return flag.ErrHelp
+			}
+
+			return UploadBlobCmd(ctx, *fileRef, *ct, args[0])
+		},
+	}
+}
+
+func UploadBlobCmd(ctx context.Context, fileRef, contentType, imageRef string) error {
+
+	ref, err := name.ParseReference(imageRef)
+	if err != nil {
+		return err
+	}
+
+	b, err := ioutil.ReadFile(fileRef)
+	if err != nil {
+		return err
+	}
+
+	if contentType == "" {
+		contentType = http.DetectContentType(b)
+	}
+
+	fmt.Fprintf(os.Stderr, "Uploading file from [%s] to [%s] with media type [%s]\n", fileRef, ref.Name(), contentType)
+	img, err := cremote.Upload(b, ref, types.MediaType(contentType), authn.DefaultKeychain)
+	if err != nil {
+		return err
+	}
+	sha := sha256.Sum256(b)
+	blobURL := ref.Context().Registry.RegistryStr() + "/v2/" + ref.Context().RepositoryStr() + "/blobs/sha256:" + hex.EncodeToString(sha[:])
```

Comment:
Wish there was a better way to do this but I don't have any ideas.

Reply:
what's wrong with it, other than being a long read?

Reply:
I could whip up another set of `Reference`s for ya. It'll involve a lot of copy-paste